DigitalLearningSolutions.Data/DataServices/SelfAssessmentDataService/CompetencyDataService.cs

```diff
@@ -0,0 +1,532 @@
+            INNER JOIN CompetencyAssessmentQuestions AS CAQ
+                ON CAQ.CompetencyID = C.ID
+            INNER JOIN AssessmentQuestions AS AQ
+                ON AQ.ID = CAQ.AssessmentQuestionID
+            INNER JOIN CandidateAssessments AS CA
+                ON CA.ID = @candidateAssessmentId
+            LEFT OUTER JOIN LatestAssessmentResults AS LAR
+                ON LAR.CompetencyID = C.ID AND LAR.AssessmentQuestionID = AQ.ID
+            INNER JOIN SelfAssessmentStructure AS SAS
+                ON C.ID = SAS.CompetencyID AND SAS.SelfAssessmentID = CA.SelfAssessmentID
+            INNER JOIN CompetencyGroups AS CG
+                ON SAS.CompetencyGroupID = CG.ID AND SAS.SelfAssessmentID = CA.SelfAssessmentID
+            LEFT OUTER JOIN CandidateAssessmentOptionalCompetencies AS CAOC
+                ON CA.ID = CAOC.CandidateAssessmentID AND C.ID = CAOC.CompetencyID AND CG.ID = CAOC.CompetencyGroupID";
+
+        public IEnumerable<int> GetCompetencyIdsForSelfAssessment(int selfAssessmentId)
+        {
+            return connection.Query<int>(
+                @"SELECT
+                        CompetencyID
+                    FROM SelfAssessmentStructure
+                    WHERE SelfAssessmentID = @selfAssessmentId",
+                new { selfAssessmentId }
+            );
+        }
+
+        public Competency? GetNthCompetency(int n, int selfAssessmentId, int candidateId)
+        {
+            Competency? competencyResult = null;
+            return connection.Query<Competency, AssessmentQuestion, Competency>(
+                $@"WITH CompetencyRowNumber AS
+                    (
+                        SELECT
+                            DENSE_RANK() OVER (ORDER BY SAS.Ordering) as RowNo,
+                            sas.CompetencyID
+                        FROM SelfAssessmentStructure as sas
+                        INNER JOIN CandidateAssessments AS CA
+                            ON CA.SelfAssessmentID = @selfAssessmentId AND CA.CandidateID = @candidateId
+                        LEFT OUTER JOIN CandidateAssessmentOptionalCompetencies AS CAOC
+                            ON CA.ID = CAOC.CandidateAssessmentID AND sas.CompetencyID = CAOC.CompetencyID AND sas.CompetencyGroupID = CAOC.CompetencyGroupID
+                        WHERE (sas.SelfAssessmentID = @selfAssessmentId) AND ((sas.Optional = 0) OR (CAOC.IncludedInSelfAssessment = 1))
+                    ),
+                    {LatestAssessmentResults}
+                    SELECT {CompetencyFields}
+                    FROM {CompetencyTables}
+                    INNER JOIN CompetencyRowNumber AS CRN
+                        ON CRN.CompetencyID = C.ID
+                    WHERE (CAOC.IncludedInSelfAssessment = 1 OR SAS.Optional = 0) AND CRN.RowNo = @n
+                    ORDER BY SAS.Ordering, CAQ.Ordering",
+                (competency, assessmentQuestion) =>
+                {
+                    if (competencyResult == null)

```

Comment:
when will competencyResult not be null?

Reply:
I think you changed this to use a ??= somewhere else, can we do that here?

Reply:
Ah yeah, that was a Resharper recommended thing. Didn't notice this one.